File: src/appl/lminfl.f
c-----------------------------------------------------------------------
c
c  R : A Computer Language for Statistical Data Analysis
c  Copyright (C) 1996, 1997  Robert Gentleman and Ross Ihaka
c  Copyright (C) 2003-5      The R Foundation
c
c  This program is free software; you can redistribute it and/or modify
c  it under the terms of the GNU General Public License as published by
c  the Free Software Foundation; either version 2 of the License, or
c  (at your option) any later version.
c
c  This program is distributed in the hope that it will be useful,
c  but WITHOUT ANY WARRANTY; without even the implied warranty of
c  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
c  GNU General Public License for more details.
c
c  You should have received a copy of the GNU General Public License
c  along with this program; if not, write to the Free Software
c  Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
c
c-----------------------------------------------------------------------
c
c   lminfl computes basic quantities useful for computing
c   regression diagnostics.
c
c   on entry
c
c       x       double precision(ldx,k)
c               the qr decomposition as computed by dqrdc or dqrdc2.
c
c       ldx     integer
c               the leading dimension of the array x.
c
c       n       integer
c               the number of rows of the matrix x.
c
c       k       integer
c               the number of columns in the matrix k.
c
c	docoef  integer (logical) indicating if  coef(*,*) should be computed
c		Computation of coef(.) is O(n^2 * k) which may be too much.
c
c       qraux   double precision(k)
c               auxiliary information about the qr decomposition.
c
c       resid   double precision(k)
c               the residuals from the regression.
c
c   on return
c
c       hat     double precision(n)
c               the diagonal of the hat matrix.
c
c       coef    double precision(n,p)
c               a matrix which has as i-th row the estimated
c               regression coefficients when the i-th case is omitted
c               from the regression.
c
c       sigma   double precision(n)
c               the i-th element of sigma contains an estimate
c               of the residual standard deviation for the model with
c               the i-th case omitted.
c
c   This version dated Aug 24, 1996.
c   Ross Ihaka, University of Auckland.
c   `docoef' option added Feb 17, 2003;  Martin Maechler ETH Zurich.
c   Handle hat == 1 case, Nov 2005.

      subroutine lminfl(x, ldx, n, k, docoef, qraux, resid,
     +     hat, coef, sigma, tol)
      integer ldx, n, k, docoef
      double precision x(ldx,k), qraux(k), resid(n),
     +     hat(n), coef(n,k), sigma(n)
c   coef(.,.) can be dummy(1) when docoef is 0(false)

      integer i, j, info
      double precision sum, denom, dummy
c
c     hat matrix diagonal
c
      do 10 i = 1,n
        hat(i) = 0.0d0
   10 continue

      do 40 j = 1,k
        do 20 i = 1,n
          sigma(i) = 0.0d0
   20   continue
        sigma(j) = 1.0d0
        call dqrsl(x, ldx, n, k, qraux, sigma, sigma, dummy,
     .       dummy, dummy, dummy, 10000, info)
        do 30 i = 1, n
          hat(i) = hat(i)+sigma(i)*sigma(i)
   30   continue
   40 continue
      do 45 i = 1, n
        if(hat(i) .ge. 1.0d0 - tol) hat(i) = 1.0d0
   45 continue
c
c     changes in the estimated coefficients
c
      if(docoef .ne. 0) then
         do 70 i = 1,n
            do 50 j = 1,n
               sigma(j) = 0.0d0
   50       continue
c           if hat is effectively 1, change is zero
            if(hat(i) .lt. 1.0d0) then
               sigma(i) = resid(i)/(1.0d0 - hat(i))
               call dqrsl(x, ldx, n, k, qraux, sigma, dummy, sigma,
     .                    dummy, dummy, dummy, 1000, info)
               call dtrsl(x, ldx, k, sigma, 1, info)
            endif
            do 60 j = 1,k
               coef(i,j) = sigma(j)
   60       continue
   70    continue
      endif
c
c     estimated residual standard deviation
c
      denom = (n - k - 1)
      sum = 0.0d0
      do 80 i = 1,n
        sum = sum + resid(i)*resid(i)
   80 continue
      do 90 i = 1,n
        if(hat(i) .lt. 1.0d0) then 
           sigma(i) = sqrt((sum-resid(i)*resid(i)/(1.0d0-hat(i)))/denom)
        else
           sigma(i) = sqrt(sum/denom)
        endif
   90 continue
      return
      end

File: src/appl.c/lminfl.c
/* -----------------------------------------------------------------------
 *
 *  R : A Computer Langage for Statistical Data Analysis
 *  Copyright (C) 1996, 1997  Robert Gentleman and Ross Ihaka
 *
 *  This program is free software; you can redistribute it and/or modify
 *  it under the terms of the GNU General Public License as published by
 *  the Free Software Foundation; either version 2 of the License, or
 *  (at your option) any later version.
 *
 *  This program is distributed in the hope that it will be useful,
 *  but WITHOUT ANY WARRANTY; without even the implied warranty of
 *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *  GNU General Public License for more details.
 *
 *  You should have received a copy of the GNU General Public License
 *  along with this program; if not, write to the Free Software
 *  Foundation, Inc., 675 Mass Ave, Cambridge, MA 02139, USA.
 *
 * -----------------------------------------------------------------------
 *
 *     lminfl computes basic quantities useful for computing
 *     regression diagnostics.
 *
 *     on entry
 *
 *         x         double precision(ldx,k)
 *                   the qr decomposition as computed by dqrdc or dqrdc2.
 *
 *         ldx       int
 *                   the leading dimension of the array x.
 *
 *         n         int
 *                   the number of rows of the matrix x.
 *
 *         k         int
 *                   the number of columns in the matrix k.
 *
 *         qraux     double precision(k)
 *                   auxiliary information about the qr decomposition.
 *
 *         b         double precision(k)
 *                   the least-squares parameter estimates.
 *
 *         resid     double precision(k)
 *                   the residuals from the regression.
 *
 *     on return
 *
 *         hat       double precision(n)
 *                   the diagonal of the hat matrix.
 *
 *         coef      double precision(n,p)
 *                   a matrix which has as i-th row contains the estimated
 *                   regression coefficients when the i-th case is omitted
 *                   from the regression.
 *
 *         sigma     double precision(n)
 *                   the i-th element of sigma contains an estimate
 *                   of the residual standard deviation for the model with
 *                   the i-th case omitted.
 *
 *     This version dated Aug 24, 1996.
 *     Ross Ihaka, University of Auckland.
 */

#include "Fortran.h"
#include "Blas.h"
#include "Linpack.h"

static int c__10000 = 10000;
static int c__1000 = 1000;
static int c__1 = 1;

int 
lminfl_(double *x, int *ldx, int *n, int
	*k, double *qraux, double *b, double *resid, double *
	hat, double *coef, double *sigma)
{
	double denom;
	double dummy, sum;
	double sqrt();
/*
	extern int F77_SYMBOL(dqrsl) (), F77_SYMBOL(dtrsl) ();
*/
	int info, i, j;
	int x_dim1, x_offset, coef_dim1, coef_offset, i__1, i__2;

	--sigma;
	--hat;
	--resid;
	coef_dim1 = *n;
	coef_offset = coef_dim1 + 1;
	coef -= coef_offset;
	--b;
	--qraux;
	x_dim1 = *ldx;
	x_offset = x_dim1 + 1;
	x -= x_offset;

	/* hat matrix diagonal */

	i__1 = *n;
	for (i = 1; i <= i__1; ++i) {
		hat[i] = 0.;
	}
	i__1 = *k;
	for (j = 1; j <= i__1; ++j) {
		i__2 = *n;
		for (i = 1; i <= i__2; ++i) {
			sigma[i] = 0.;
		}
		sigma[j] = 1.;
		F77_SYMBOL(dqrsl) (&x[x_offset], ldx, n, k, &qraux[1], &sigma[1], &sigma[1], &
			   dummy, &dummy, &dummy, &dummy, &c__10000, &info);
		i__2 = *n;
		for (i = 1; i <= i__2; ++i) {
			hat[i] += sigma[i] * sigma[i];
		}
	}

	/* changes in the estimated coefficients */

	i__1 = *n;
	for (i = 1; i <= i__1; ++i) {
		i__2 = *n;
		for (j = 1; j <= i__2; ++j) {
			sigma[j] = 0.;
		}
		sigma[i] = resid[i] / (1. - hat[i]);
		F77_SYMBOL(dqrsl) (&x[x_offset], ldx, n, k, &qraux[1], &sigma[1], &dummy, &sigma[
			       1], &dummy, &dummy, &dummy, &c__1000, &info);
		F77_SYMBOL(dtrsl) (&x[x_offset], ldx, k, &sigma[1], &c__1, &info);
		i__2 = *k;
		for (j = 1; j <= i__2; ++j) {
			coef[i + j * coef_dim1] = sigma[j];
		}
	}

	/* estimated residual standard deviation */

	denom = (double) (*n - *k - 1);
	sum = 0.;
	i__1 = *n;
	for (i = 1; i <= i__1; ++i) {
		sum += resid[i] * resid[i];
	}
	i__1 = *n;
	for (i = 1; i <= i__1; ++i) {
		sigma[i] = sqrt((sum - resid[i] * resid[i] / (1. - hat[i])) / denom);
	}
	return 0;
}
